### PR TITLE
New: `require-await` rule (fixes #6820)

### LINKED ISSUE #6820: Rule Proposal: require-await
Async functions and await expressions arrived stage 4.
We can add the rule for those after our parser supported those.

This proposal is the `await` version of [require-yield](http://eslint.org/docs/rules/require-yield) rule.
It disallows async functions without await expressions.

``` js
/*eslint require-await: error*/

//------------------------------------------------------------------------------
// ✔ GOOD
async function foo() {
    await doSomething()
}

async function foo() {
    // do nothing.
}

//------------------------------------------------------------------------------
// ✘ BAD
async function foo() {
    doSomething()
}
```

### PR DESCRIPTION
**Pull request by mysticatea**
**What is the purpose of this pull request? (put an "X" next to item)**

[X] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))

**Please describe what the rule should do:**

This rule disallows `async` functions which have no `await` expression.
Async functions which have no `await` expression may be the unintentional result of refactoring.

See #6820 also.

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

``` js
/*eslint require-await: error*/

//------------------------------------------------------------------------------
// ✔ GOOD
async function foo() {
    await doSomething()
}

// This rule ignores empty functions as same as `require-yield` rule.
async function noop() {
    // do nothing.
}

function foo() {
    doSomething()
}

//------------------------------------------------------------------------------
// ✘ BAD
async function foo() {
    doSomething()
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
- This is a general matter for JavaScript language.
- ESLint has `require-yield` rule which does the same check for generator functions.

**What changes did you make? (Give an overview)**

This PR adds the `require-await` rule into core.

In addition, this PR adds 2 functions to `ast-utils`.
- `getFunctionNameWithKind(node)` gets the name and kind of the given function. I intended that this is an utility to use for #7260. Please see JSDoc comment for details.
- `getFunctionHeadLoc(node, sourceCode)` gets the location of the function head. I intended that this is an utility to use for #3307. Please see JSDoc comment for details.

**Is there anything you'd like reviewers to focus on?**
- Messages and documents.
- Added 2 utilities.

### CODE REVIEW

File: lib/rules/require-await.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/**
+ * @fileoverview Rule to
```
Comment (platinumazure) [inline lib/rules/require-await.js]: This needs to be fleshed out, I think. :smile:

File: lib/rules/require-await.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Checks whether the given function node is an empty function or not.
+ *
+ * @param {ASTNode} node - The function node to check.
+ * @returns {boolean} `true` if the function node is an empty function.
+ */
+function isEmpty(node) {
```
Comment (platinumazure) [inline lib/rules/require-await.js]: Seems like this should go in ast-utils?

File: lib/rules/require-await.js (hunk 1 above)
Comment (mysticatea) [inline lib/rules/require-await.js]: Oops, I have forgotten it since I use this template after a long time.

File: lib/rules/require-await.js (hunk 2 above)
Comment (mysticatea) [inline lib/rules/require-await.js]: Probably.

**Comment by nzakas** [comment]
Just a heads up: we have moved to a new CLA checker on pull requests. Even if you've previously signed our CLA, we will need to you sign the new one. To do so, look at the status checks for **licence/cla** and click the "Details" link. Sorry for the inconvenience.

**Comment by vitorbal** [comment]
@platinumazure I believe your concerns have been addressed in the new commits from @mysticatea . Could you take a look when you have a chance?

**Pull request merged**